File: backend/dotnet/src/WorkApi/Repositories/TaskRepository.cs
using MongoDB.Driver;
using WorkApi.Models;

namespace WorkApi.Repositories;

public class TaskRepository : RepositoryBase<TaskModel>, ITaskRepository
{
    public TaskRepository(IMongoDatabase database) : base(database, "tasks")
    {
    }

    public Task<List<TaskModel>> GetByProjectIdAsync(string projectId, int skip = 0, int limit = 100)
    {
        var filter = Builders<TaskModel>.Filter.Eq(x => x.ProjectId, projectId);
        return FindAsync(filter, skip, limit);
    }

    public Task<List<TaskModel>> GetByStatusAsync(string status, int skip = 0, int limit = 20)
    {
        var filter = Builders<TaskModel>.Filter.Eq(x => x.Status, status);
        return FindAsync(filter, skip, limit);
    }

    public Task<List<TaskModel>> GetByAssignedUserIdAsync(string assignedUserId, int skip = 0, int limit = 20)
    {
        var filter = Builders<TaskModel>.Filter.Eq(x => x.AssignedUserId, assignedUserId);
        return FindAsync(filter, skip, limit);
    }
}

File: backend/node/src/repositories/task.repository.ts
import { Db, Filter, ObjectId } from 'mongodb';
import { Task, TaskStatus } from '../models/entities';
import { BaseRepository } from './base.repository';

export interface TaskFilter {
  projectId?: ObjectId;
  status?: TaskStatus;
}

export class TaskRepository extends BaseRepository<Task> {
  constructor(db: Db) {
    super(db, 'tasks');
  }

  async findByProjectId(projectId: ObjectId): Promise<Task[]> {
    return this.collection.find({ projectId }).toArray();
  }

  async findFiltered(filter: TaskFilter, limit = 50, offset = 0): Promise<Task[]> {
    const query: Filter<Task> = {};
    if (filter.projectId) query.projectId = filter.projectId;
    if (filter.status) query.status = filter.status;
    return this.collection.find(query).skip(offset).limit(limit).toArray();
  }

  async countFiltered(filter: TaskFilter): Promise<number> {
    const query: Filter<Task> = {};
    if (filter.projectId) query.projectId = filter.projectId;
    if (filter.status) query.status = filter.status;
    return this.collection.countDocuments(query);
  }
}
